x/autopilot/keeper/ibc.go

```diff
@@ -0,0 +1,162 @@
+	if !ok {
+		return coin, fmt.Errorf("unable to parse amount from transfer packet: %v", transferMetadata)
+	}
+	coin = sdk.NewCoin(transferMetadata.Denom, amount)
+	return coin, nil
+}
+
+// In the event of an ack error after a outbound transfer, we'll have to bank send to a fallback address
+func (k Keeper) SendToFallbackAddress(ctx sdk.Context, packetData []byte, fallbackAddress string) error {
+	// First unmarshal the transfer metadata to get the sender/reciever, and token amount/denom
+	var transferMetadata transfertypes.FungibleTokenPacketData
+	if err := transfertypes.ModuleCdc.UnmarshalJSON(packetData, &transferMetadata); err != nil {
+		return err
+	}
+
+	// Pull out the original sender of the transfer which will also be the bank sender
+	sender := transferMetadata.Sender
+	senderAccount, err := sdk.AccAddressFromBech32(sender)
+	if err != nil {
+		return errorsmod.Wrapf(err, "invalid sender address")
+	}
+	fallbackAccount, err := sdk.AccAddressFromBech32(fallbackAddress)
+	if err != nil {
+		return errorsmod.Wrapf(err, "invalid fallback address")
+	}
+
+	// Build the token from the transfer metadata
+	token, err := k.BuildCoinFromTransferMetadata(transferMetadata)
+	if err != nil {
+		return err
+	}
+
+	// Finally send to the fallback account
+	if err := k.bankKeeper.SendCoins(ctx, senderAccount, fallbackAccount, sdk.NewCoins(token)); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// If there was a failed ack from an outbound transfer of one of the autopilot actions,
+// we'll need to check if there was a fallback address. If one was stored, bank send
+// to that fallback address
+// If the ack was successful, we should delete the address (if it exists)
+func (k Keeper) OnAcknowledgementPacket(ctx sdk.Context, packet channeltypes.Packet, acknowledgement []byte) error {
+	// Retrieve the fallback address for the given packet
+	// We use the packet source channel here since this will correspond with the channel on Stride
+	channelId := packet.SourceChannel
+	sequence := packet.Sequence
+	fallbackAddress, fallbackAddressFound := k.GetTransferFallbackAddress(ctx, channelId, sequence)
+
+	// If there was no fallback address, there's nothing else to do
+	if !fallbackAddressFound {
+		return nil
+	}
+
+	// Remove the fallback address since the packet is no longer pending
+	k.RemoveTransferFallbackAddress(ctx, channelId, sequence)
+
+	// Check whether the ack was successful or was an ack error
+	success, err := k.CheckAcknowledgementStatus(ctx, acknowledgement)
+	if err != nil {
+		return err
+	}
+
+	// If successful, no additional action is necessary
+	if success {
+		return nil
+	}
+
+	// If the ack was an error, we'll need to bank send to the fallback address
+	return k.SendToFallbackAddress(ctx, packet.Data, fallbackAddress)
+}
+
+// If there's a timed out packet, we'll infinitely retry the transfer
+func (k Keeper) OnTimeoutPacket(ctx sdk.Context, packet channeltypes.Packet) error {
```

Comment:
Hmm, I'm not sure I see the attack vector here? I think traditionally the attack would be to send a message with a short timeout such that it always times-out and infinitely retries

But we set the timeout on chain so there's no way to really "force" a timeout (to my understanding). And again, this is different than an ack failure (which you can force)

Reply:
I'm not opposed to capping the retries - but I don't think this functionality increases the potential for spam. I could be wrong though!

Reply:
was thinking something along these lines
* spin up newchain, but don't expose the rpc
* create a client for newchain on stride, as well as a connection/transfer channel (e.g. channel-N)
* start ls + forwarding small token amounts along channel-N - for example if the timeout is 1 min, fill blocks with these txs for 1 min
* now, there's a backlog of txs which will take 1 min to clear out on stride (but they will all time out in 1 min). as the attacker, you have access to the rpc so can force the timeouts 

I take back the point about relayers running out of funds, don't think that would happen here (since it's just the attacker's relayer)

Reply:
ah that's an interesting point! 

Although, to be clear, you wouldn't be able to set the timeout to 1 minute. So I think the attack would be that you:
* send a bunch of txs
* turn off you relayer until they timeout
* then turn it back on so they timeout and retry (probably forking the relayer so that it only relays one message type at a time, otherwise the timeout + retry will both succeed)
* then turn it off again so they timeout again 
* repeat

Nonetheless, I agree, I think it's best to avoid the infinite loop to be safe!

Reply:
yeah, the 1 min is just an example (if we set it to 10 min or 30 min, the attacker would need to fill blocks for that duration). 

although maybe this attack vector is equivalent to someone just filling blocks with standard txs (there's nothing special about them coming from the relayer)

slightly tricky to reason about - would be curious to hear @riley-stride @shellvish thoughts on this as well! 

if there's no attack vector, it's definitely simpler / less code to use the approach with infinite retries

Reply:
Yeah I think considering they'd be running the relayer themselves, and they'd have to spend gas on stride each time they process a timeout, I'm not sure there'd be an advantage to this approach versus standard txs.

But the infinite nature does make me a little worried. If there was a way to pay for X transactions and be able to actually submit X * infinity, that would be very bad. (doesn't seem like there is right now, but possible there's one we're not thinking of)

Reply:
Interesting idea.

I see two potentially-blockspace-consuming operations here. 
1) Epochly IBC-transfers from the begin blocker (initiated by stride protocol every timeout period)
2) Relayer relaying those IBC-transfers

#2 seems like a non-issue (relaying these is no different than a standard tx) per Sam's logic above:
> I think considering they'd be running the relayer themselves, and they'd have to spend gas on stride each time they process a timeout, I'm not sure there'd be an advantage to this approach versus standard txs.

#1 could congest the network _if_ beginblocker ops are blockspace-consuming, but I don't _think_ they are (correct me please).

If #1 is in fact an issue (beginblocker ops do consume blockspace) one creative approach could be to whitelist which chains can be LS+forwarded to. We'd set a pretty broad list in the upgrade handler and allow adding to it through governance.

Reply:
Discussed offline and decided on:
*  long timeout (3 hours) 
* no retries
* fallback on timeout

Rationale being that:
* We want the timeout to be long enough to handle a stressed environment (empirically we've found that 2 hours has been long enough in the past for significant moments)
* We want to avoid infinite retries due to the unknown
* Retries don't seem to add much value (the relayers are either on and are relaying, or their off and their not - retrying doesn't increase the likelihood that a message is relayed)